app/scripts/metamask-controller.test.js

```diff
@@ -1273,4 +1294,202 @@ describe('MetaMaskController', function () {
+        { ...newNetworkConfiguration, id: networkConfigurationId },
+      );
+    });
+
+    it('adds new networkConfiguration to networkController store and sets it as current provider if options.setActive is true', async function () {
+      const newNetworkConfiguration = {
+        rpcUrl: 'https://new-chain-rpc-url',
+        chainId: '0x9999',
+        ticker: 'NEW',
+        chainName: 'new-chain',
+        rpcPrefs: {},
+      };
+      const networkConfigurationId =
+        await metamaskController.upsertNetworkConfiguration(
+          newNetworkConfiguration,
+          { setActive: true },
+        );
+
+      assert.deepStrictEqual(
+        metamaskController.networkController.store.getState()
+          .networkConfigurations[networkConfigurationId],
+        { ...newNetworkConfiguration, id: networkConfigurationId },
+      );
+
+      assert.deepStrictEqual(
+        metamaskController.networkController.store.getState().provider,
+        {
+          ...newNetworkConfiguration,
+          id: networkConfigurationId,
+          type: NETWORK_TYPES.RPC,
+        },
+      );
+    });
+
+    it('adds new networkConfiguration to networkController store and calls to the metametrics event tracking with the correct values', async function () {
+      sinon.spy(metamaskController.metaMetricsController, 'trackEvent');
+
+      const newNetworkConfiguration = {
+        rpcUrl: 'https://new-chain-rpc-url',
+        chainId: '0x9999',
+        ticker: 'NEW',
+        chainName: 'new-chain',
+        rpcPrefs: {},
+      };
+
+      await metamaskController.upsertNetworkConfiguration(
+        newNetworkConfiguration,
+        { source: 'https://chainlist.org/' },
+      );
+
+      assert(
+        metamaskController.metaMetricsController.trackEvent.calledWith({
+          event: 'Custom Network Added',
+          category: 'Network',
+          referrer: {
+            url: 'https://chainlist.org/',
+          },
+          properties: {
+            chain_id: '0x9999',
+            symbol: 'NEW',
+            source: 'dapp',
+          },
+        }),
+      );
+    });
+
+    it('adds new networkConfiguration to networkController store, but only adds valid properties (rpcUrl, chainId, ticker, chainName, rpcPrefs) and fills any missing properties from this list as undefined', async function () {
+      const networkConfigurationId =
+        await metamaskController.upsertNetworkConfiguration({
+          rpcUrl: 'https://new-chain-rpc-url',
+          chainId: '0x9999',
+          ticker: 'NEW',
+          invalidKey: 'new-chain',
+          invalidKey2: {},
+        });
+      assert.deepStrictEqual(
+        metamaskController.networkController.store.getState()
+          .networkConfigurations[networkConfigurationId],
+        {
+          rpcUrl: 'https://new-chain-rpc-url',
+          chainId: '0x9999',
+          ticker: 'NEW',
+          chainName: undefined,
+          rpcPrefs: undefined,
```

Comment:
Hmm. Not ideal that we've allowed the state to contradict the type we're using for it. The `rpcPrefs` property is marked as required in `NetworkConfiguration`.

The method being replaced by `upsertNetworkConfiguration` is `setCustomRpc`, and that used to default `rpcPrefs` to an empty object if it wasn't provided.

Reply:
right whoops! I think the `NetworkConfiguration` type should have it as undefined. I've changed it to that here: https://github.com/MetaMask/metamask-extension/pull/17421/files#diff-b3422fb8c3cd155660c8689ba9fdd32ca0dcfe799c152a10be2b8c05d47310ceR39

Or do you think it should default to an empty object as before?

Reply:
This seems OK to me!